splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/sequence/AbstractSequence.java

```diff
@@ -63,12 +77,16 @@ public long peekAtCurrentValue() throws StandardException {
 
     public abstract void close() throws IOException;
 
+    // must be called with acquired read lock
     private void allocateBlock(boolean peek) throws StandardException{
         boolean success=false;
         long absIncrement = incrementSteps < 0 ? -incrementSteps :
                                                   incrementSteps;
         while(!success){
-            updateLock.lock();
+            if(remaining.getAndDecrement()>0)
+                return;
+            rwLock.readLock().unlock();
```

Comment:
I wanted to get more understanding on why we can't simply upgrade a read lock to write lock, i.e. why the `readlock().unlock()` is necessary, I came across [this SO thread ](https://stackoverflow.com/questions/464784/java-reentrantreadwritelocks-how-to-safely-acquire-write-lock)which made me worried this might block forever? Did you double-check that?

Reply:
It is not possible to directly upgrading a read-write lock without unlocking the read lock first. I think this is common behavior not only in Java but kind of any read-write lock implementations. The problem is that multiple threads could try to upgrade simultaneously, making the scenario very complex to be correct. Downgrading is simple because the current thread knows it's the only thread doing so.

For the behavior described in the SO thread, I wouldn't worry too much until we actually see it. It looks like a bug in Java that happens rarely. If this problem happens very often, `ReentrantReadWriteLock` implementation is basically buggy and unusable, which doesn't see to be the case in practice.